src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs

```diff
@@ -199,107 +293,239 @@ public async Task<IActionResult> ExternalLoginCallback(string returnUrl = null,
+                            {
+                                var identityResult = await _signInManager.UserManager.AddLoginAsync(user, new UserLoginInfo(info.LoginProvider, info.ProviderKey, info.ProviderDisplayName));
+                                if (identityResult.Succeeded)
+                                {
+                                    _logger.LogInformation(3, "User account linked to {Name} provider.", info.LoginProvider);
+
+                                    // We have created/linked to the local user, so we must verify the login.
+                                    // If it does not succeed, the user is not allowed to login
+                                    var signInResult = await _signInManager.ExternalLoginSignInAsync(info.LoginProvider, info.ProviderKey, isPersistent: false, bypassTwoFactor: true);
+                                    if (signInResult.Succeeded)
+                                    {
+                                        return RedirectToLocal(returnUrl);
+                                    }
+                                    else
+                                    {
+                                        ModelState.AddModelError(string.Empty, T["Invalid login attempt."]);
+                                        return View(nameof(Login));
+                                    }
+                                }
+                                AddIdentityErrors(identityResult);
+                            }
+                        }
+                        return View("RegisterExternalLogin", externalLoginViewModel);
+                    }
+                }
             }
+            return RedirectToAction(nameof(Login));
         }
 
         [HttpPost]
         [AllowAnonymous]
         [ValidateAntiForgeryToken]
-        public async Task<IActionResult> ExternalLoginConfirmation(ExternalLoginViewModel model, string returnUrl = null, string loginProvider = null)
+        public async Task<IActionResult> RegisterExternalLogin(RegisterExternalLoginViewModel model, string returnUrl = null)
         {
-            if ((!model.IsExistingUser && !(await _siteService.GetSiteSettingsAsync()).As<RegistrationSettings>().UsersCanRegister))
+            IUser user = null;
+            var settings = (await _siteService.GetSiteSettingsAsync()).As<RegistrationSettings>();
+            var info = await _signInManager.GetExternalLoginInfoAsync();
+
+            if (info == null)
+            {
+                _logger.LogWarning("Error loading external login info. {externalLoginViewModel}", model);
+                return NotFound();
+            }
+
+            if (settings.UsersCanRegister == RegistrationSettings.UsersCanRegisterEnum.NoRegistration)
             {
-                _logger.LogInformation("Site does not allow user registration.");
+                _logger.LogWarning("Site does not allow user registration. {externalLoginViewModel}", model);
```

Comment:
We should not log it this way as it contains private information. Better pick the field that makes sense.

Reply:
There are other places where the view model is passed in the log btw.